### ISSUE TITLE
TextEditor: Accessibility Issues

### ISSUE THREAD

**Issue by kbrown2017** [primefaces/primefaces#7600]
The Text Editor has the following issues with accessibility by keyboard users and use of a screen reader. 

- 1.1.1, 1.3.1, 3.3.2:  Each of the <button> controls, text area of the editor, and modal textboxes are missing labels.  These labels must be added to assist the screen reader user in understanding the context of each control.  Please update each of the buttons to include an ARIA label that provides proper context.  Please reference the following URL for further guidance: https://www.w3.org/TR/WCAG20-TECHS/ARIA14.html

1. ButtonTextAreaModalTextboxesMissingARIALabel.png

- 1.3.1:  Each of the “links” found in each of the dialogs after the selection of the “link” button are only coded as <a> tags using class information as the text.  The text of “save” or “edit/delete” is not announced or tab-able using both the keyboard and screen reader.

1. OddLinkMarkup.png

- 4.1.2:  The main text area of the editor is missing an accessible role attribute.  This will assist the screen reader in announcing the type of control when navigated to.  Please update the text area to include the role of “textbox”.  Also, please reference the following URL for further guidance since this control is acting more like a text area would need to include the multiline functionality: https://www.w3.org/TR/wai-aria/states_and_properties

ARIARoleMissing.png

- 4.1.2:  The buttons, text area, and modal textboxes are missing accessible names.  These names assist the screen reader in properly announcing each of the control effectively.

AccessibleNamesMissing.png

- 4.1.2:  Each of the buttons within the control are missing the ARIA pressed attribute.  This will assist the user in understanding when one of the controls has been pressed when navigating back to the control.  Please reference the following URL for further guidance: https://www.w3.org/TR/wai-aria/states_and_properties#aria-pressed 

1. ARIAPressedAttributeMissing.png

- 4.1.2:  The link button is missing the ARIA haspopup attribute.  This will assist the screen reader in understanding that after selection of the control that there is a popup modal.  Please reference the following URL: https://www.w3.org/TR/wai-aria/states_and_properties#aria-haspopup

1. ARIAHasPopupAttributeMissing.png

- 1.4.3:  The selected state of the button is only conveyed via the use of color (changes from black to blue).  Also the color contrast of the control is only 1.6:1 to the associated button controls.  Please update the control’s color contrast to be at least 3.0:1 between the blue and black color.  Also, update the control with a visual queue to assist the color blind user in understanding that the control is indeed selected (e.g. maybe use a solid box around the control).

ColorContrastandColorOnlyUsedAfterSelection.png

- 2.4.6:  Each of the dialogs are missing proper labels.  Please update each of the modals to include a proper label (e.g. Create Link or Edit Link).

1. LabelMissingFromDialog.png

- 2.1.1, 2.4.6, 4.1.2:  Each of the modal dialogs are only coded as divs and are not keyboard or screen reader accessible.  Each of the modals must include the following:

1. A defined label for each modal dialog
2. Dialogs can either be modal or non-modal. This can be semantically defined using the aria-modal property where "true" means that it is modal and "false" means that it is not modal.
3. Set focus gets first heading within the modal dialog (similar to a 'Skip to Main Content' link), which is typically a <h1> heading (this will need to be added since they currently are missing heading markup).
4. NVDA screen reader will not read the contents of a modal dialog unless the contents are housed inside of a document (i.e. role="document"). This document role is often included as the immediate child element of the element containing the role="dialog".
5. Make sure that all content behind the modal once opened is not accessible or perceived by the screen reader application.
Pressing the Escape key within the modal dialog dismisses it and returns focus to the element that triggered the dialog to display.
IssuesWithModals.png

- 2.1.1, Rev508 FPC 302.1: After creating a link in the text editor, the screen reader user does not perceive it at all.  It seems that there is an <a> tag created but the user is unable to navigate and use the link.  Research will be needed.  Also, the keyboard only user is unable to tab to it and select it.

1. LinkNotPerceivedByScreenReaderAndIsUnTabable.png

- 2.4.7:  Visual focus is lost when navigating through each of the editor buttons.  It seems that there are tab stops assigned to not only the button controls but also to svg image associated to it.  Please update the tab order and remove all tab stops to each of the svg images.

1. ExtraTabStops.png

Each of the “dropdown” controls are missing proper ARIA names, roles, states, and values.  Also, each of the color pallet buttons are missing proper name, role, and state values and are currently not keyboard accessible at all.  Both are currently <span> only.  Please see DropdownCurrentlyCodedAsSpan.png and ColorPalletControlsCurrentlyCodedAsSpan.png.

**Comment by melloware** [comment]
PF uses QuillJS under the covers so most of these should be reported to QuillJS: https://quilljs.com/

**Comment by melloware** [comment]
For example: https://github.com/quilljs/quill/issues?q=is%3Aissue+is%3Aopen+aria
